tracks/rust/exercises/circular-buffer/mentoring.md

```diff
@@ -0,0 +1,204 @@
+
+```rust
+pub struct CircularBuffer<T> {
+    buffer: Vec<Option<T>>,
+    start: usize, // read index
+    end: usize,   // write index
+}
+
+#[derive(Debug, PartialEq)]
+pub enum Error {
+    EmptyBuffer,
+    FullBuffer,
+}
+
+impl<T> CircularBuffer<T> {
+    pub fn new(capacity: usize) -> Self {
+        CircularBuffer {
+            buffer: (0..capacity).map(|_| None).collect(),
+            start: 0,
+            end: 0,
+        }
+    }
+
+    pub fn write(&mut self, element: T) -> Result<(), Error> {
+        match self.buffer[self.end] {
+            Some(_) => Err(Error::FullBuffer),
+            None => {
+                self.buffer[self.start] = Some(element);
+                self.advance_end();
+                Ok(())
+            }
+        }
+    }
+
+    pub fn read(&mut self) -> Result<T, Error> {
+        self.buffer[self.start]
+            .take()
+            .ok_or(Error::EmptyBuffer)
+            .map(|elem| {
+                self.advance_start();
+                elem
+            })
+    }
+
+    pub fn clear(&mut self) {
+        self.start = 0;
+        self.end = 0;
+
+        self.buffer.iter_mut().for_each(|elem| *elem = None);
+    }
+
+    pub fn overwrite(&mut self, element: T) {
+        if self.buffer[self.start].is_some() {
+            self.advance_start();
+        }
+
+        self.buffer[self.end] = Some(element);
+        self.advance_end();
+    }
+
+    fn advance_start(&mut self) {
+        self.start = (self.start + 1) % self.buffer.len();
+    }
+
+    fn advance_end(&mut self) {
+        self.end = (self.end + 1) % self.buffer.len();
+    }
+}
+
+```
+
+### Common Suggestions
+
+- Usage of `unwrap` should be as minimal as possible, preferably non-existent if we can help it. Take the following implementation of the `read` function that calls `unwrap`:
+```rust
+fn read(&mut self) -> Result<T, Error> {
+  match &self.buffer[self.read_index] {
+    Some(_) => {
+      let old_index = self.read_index;
+      self.update_read_index();
+      Ok(self.buffer[old_index].take().unwrap())
+    },
+    None => Err(Error::EmptyBuffer),
+  }
+}
+```
+
+How can we eliminate the `unwrap` call on `take`? The handy `ok_or` method is a prime candidate. This implementation can be restructured into something like this:
+```rust
+fn read(&mut self) -> Result<T, Error> {
+  self.buffer[self.read_index]
+    .take()
+    .ok_or(Error::EmptyBuffer)
+    .map(|element| {
+      self.update_read_index();
+      element
+    })
+}
+```
+
+While the first implementation is unlikely to ever actually panic due to the `unwrap` call due to how it's structured, imparting strategies for how to limit usage of `unwrap` will certainly be beneficial to the student(s). 
+
+- Logically distinct operations such as advancing the read index and advancing the write index should be relegated to their own functions to help with code readability, organization, and to adhere to the DRY principle. 
```

Comment:
List is not correctly formatted (all two spaces), but I would suggest not using a list at all but add subheaders.